tests/bundle-tests/testem.js

```diff
@@ -0,0 +1,24 @@
+/* eslint-env node */
+"use strict";
+var fluid = require("../../src/module/fluid");
+fluid.setLogging(true);
+
+require("gpii-testem");
+
+fluid.defaults("fluid.tests.bundle.testem", {
+    gradeNames: ["gpii.testem.coverage"],
+    testPages:  ["tests/bundle-tests/all-bundle-tests.html"],
+    sourceDirs: {
+        src: "%infusion/src"
+    },
+    contentDirs: {
+        tests:   "%infusion/tests"
+    },
+    testemOptions: {
+        skip: "PhantomJS,Opera,Safari",
```

Comment:
For some reason this ends up running in Chrome twice, once in a headless mode and then again in the real browser. It seems unnecessary to test this cross-platform, is there a configuration we can apply to skip all browsers except headless Chrome, i.e. skip IE, Firefox and Chrome proper itself? What is the name of the headless Chrome launcher?
However, what is disturbing is that there is evidence that not all the tests actually run in all browsers. Here is the output I get:
````
ok 1 Firefox 64.0 - [753 ms] - Bundle Tests: infusion-all.html
ok 2 Firefox 64.0 - [696 ms] - Bundle Tests: infusion-all-min.html
ok 3 Firefox 64.0 - [2 ms] - Verifying that the infusion-all-no-jquery.js bundle can be safely included by a browser.
ok 4 Firefox 64.0 - [643 ms] - Bundle Tests: infusion-all-no-jquery.html
ok 5 Firefox 64.0 - [706 ms] - Bundle Tests: infusion-all-no-jquery-min.html
ok 6 Firefox 64.0 - [1 ms] - Verifying that the infusion-framework.js bundle can be safely included by a browser.
ok 7 Firefox 64.0 - [265 ms] - Bundle Tests: infusion-framework.html
ok 8 Firefox 64.0 - [1 ms] - Verifying that the infusion-framework.min.js bundle can be safely included by a browser.
ok 9 Firefox 64.0 - [215 ms] - Bundle Tests: infusion-framework-min.html
ok 10 Firefox 64.0 - [209 ms] - Bundle Tests: infusion-framework-no-jquery.html
ok 11 Firefox 64.0 - [1 ms] - Verifying that the infusion-framework-no-jquery.min.js bundle can be safely included by a browser.
ok 12 Firefox 64.0 - [220 ms] - Bundle Tests: infusion-framework-no-jquery-min.html
ok 13 Firefox 64.0 - [567 ms] - Bundle Tests: infusion-uio.html
ok 14 Firefox 64.0 - [566 ms] - Bundle Tests: infusion-uio-min.html
ok 15 Firefox 64.0 - [631 ms] - Bundle Tests: infusion-uio-no-jquery.html
ok 16 Firefox 64.0 - [1 ms] - Verifying that the infusion-uio-no-jquery.min.js bundle can be safely included by a browser.
ok 17 Firefox 64.0 - [613 ms] - Bundle Tests: infusion-uio-no-jquery-min.html
ok 18 IE 11.0 - [5673 ms] - Bundle Tests: infusion-all.html
ok 19 IE 11.0 - [2991 ms] - Bundle Tests: infusion-all-min.html
ok 20 IE 11.0 - [2978 ms] - Bundle Tests: infusion-all-no-jquery.html
ok 21 IE 11.0 - [1596 ms] - Bundle Tests: infusion-all-no-jquery-min.html
ok 22 IE 11.0 - [994 ms] - Bundle Tests: infusion-framework.html
ok 23 IE 11.0 - [1 ms] - Verifying that the infusion-framework.min.js bundle can be safely included by a browser.
ok 24 IE 11.0 - [534 ms] - Bundle Tests: infusion-framework-min.html
ok 25 IE 11.0 - [2 ms] - Verifying that the infusion-framework-no-jquery.js bundle can be safely included by a browser.
ok 26 IE 11.0 - [459 ms] - Bundle Tests: infusion-framework-no-jquery.html
ok 27 IE 11.0 - [262 ms] - Bundle Tests: infusion-framework-no-jquery-min.html
ok 28 IE 11.0 - [3453 ms] - Bundle Tests: infusion-uio.html
ok 29 IE 11.0 - [1711 ms] - Bundle Tests: infusion-uio-min.html
ok 30 IE 11.0 - [1707 ms] - Bundle Tests: infusion-uio-no-jquery.html
ok 31 IE 11.0 - [1029 ms] - Bundle Tests: infusion-uio-no-jquery-min.html
ok 32 IE 11.0 - [2 ms] - Verifying that the infusion-uio-no-jquery.min.js bundle can be safely included by a browser.
ok 33 Chrome 71.0 - [583 ms] - Bundle Tests: infusion-all.html
ok 34 Chrome 71.0 - [538 ms] - Bundle Tests: infusion-all-min.html
ok 35 Chrome 71.0 - [529 ms] - Bundle Tests: infusion-all-no-jquery.html
ok 36 Chrome 71.0 - [486 ms] - Bundle Tests: infusion-all-no-jquery-min.html
ok 37 Chrome 71.0 - [175 ms] - Bundle Tests: infusion-framework.html
ok 38 Chrome 71.0 - [161 ms] - Bundle Tests: infusion-framework-min.html
ok 39 Chrome 71.0 - [154 ms] - Bundle Tests: infusion-framework-no-jquery.html
ok 40 Chrome 71.0 - [143 ms] - Bundle Tests: infusion-framework-no-jquery-min.html
ok 41 Chrome 71.0 - [452 ms] - Bundle Tests: infusion-uio.html
ok 42 Chrome 71.0 - [461 ms] - Bundle Tests: infusion-uio-min.html
ok 43 Chrome 71.0 - [420 ms] - Bundle Tests: infusion-uio-no-jquery.html
ok 44 Chrome 71.0 - [388 ms] - Bundle Tests: infusion-uio-no-jquery-min.html
ok 45 Chrome 71.0 - [0 ms] - Verifying that the infusion-uio-no-jquery.min.js bundle can be safely included by a browser.
ok 46 Chrome 71.0 - [588 ms] - Bundle Tests: infusion-all.html
ok 47 Chrome 71.0 - [490 ms] - Bundle Tests: infusion-all-min.html
ok 48 Chrome 71.0 - [528 ms] - Bundle Tests: infusion-all-no-jquery.html
ok 49 Chrome 71.0 - [491 ms] - Bundle Tests: infusion-all-no-jquery-min.html
ok 50 Chrome 71.0 - [190 ms] - Bundle Tests: infusion-framework.html
ok 51 Chrome 71.0 - [187 ms] - Bundle Tests: infusion-framework-min.html
ok 52 Chrome 71.0 - [141 ms] - Bundle Tests: infusion-framework-no-jquery.html
ok 53 Chrome 71.0 - [141 ms] - Bundle Tests: infusion-framework-no-jquery-min.html
ok 54 Chrome 71.0 - [440 ms] - Bundle Tests: infusion-uio.html
ok 55 Chrome 71.0 - [447 ms] - Bundle Tests: infusion-uio-min.html
ok 56 Chrome 71.0 - [378 ms] - Bundle Tests: infusion-uio-no-jquery.html
ok 57 Chrome 71.0 - [362 ms] - Bundle Tests: infusion-uio-no-jquery-min.html
ok 58 Chrome 71.0 - [0 ms] - Verifying that the infusion-uio-no-jquery.min.js bundle can be safely included by a browser.
````
Note that 17 tests run in Firefox, 15 tests run in IE 11, 27 tests run in total in Chrome.
What I am thinking is that this is evidence of a long-standing bug we have had in testem - that it cleans up browser instances before they have actually finished queuing any tests. This probably explains the random variation that I am often seeing in coverage reports. I think currently we are dealing with this with some kind of timeout which is not principled enough. We need a sound solution based on some kind of interlock, perhaps injected as a script at the end of the file indicating that all the tests which are going to be queued are now queued.

Reply:
I'll have to investigate, I'm also a bit confused about why some tests have assertion messages and others just have a summary for the whole HTML fixture.  That seems suspicious.

Reply:
OK, @amb26, I think we may actually be double-counting rather than skipping anything. 
 There are actually twelve HTML files, and there should be one summary message reported for each, for a total of 12 tests per browser.  Somehow, more than 12 tests per browser are reported. What should usually happen is that only the "Bundle Tests" (fixture-level) messages show up, but sometimes the individual tests within the rollup are counted as well.

I suspect this is related to using a QUnit.composite rollup in combination with Testem.   To test this theory, I have bypassed the bundle tests rollup altogether in the latest commit.  If this resolves the stability of the overall count between runs, then we have a clear area to investigate.

I have also configured the tests to only run in headless chrome.

Reply:
FYI, the launcher name was trivial to determine, the command is: `node node_modules/testem/testem.js launchers`